api/external_notion_test.go

```diff
@@ -0,0 +1,168 @@
+package api
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+
+	jwt "github.com/golang-jwt/jwt"
+)
+
+const (
+	notionUser           string = `{"id":"notionTestId","name":"Notion Test","picture":"http://example.com/avatar","email":"notion@example.com","verified_email":true}}`
+	notionUserWrongEmail string = `{"id":"notionTestId","name":"Notion Test","picture":"http://example.com/avatar","email":"other@example.com","verified_email":true}}`
+	notionUserNoEmail    string = `{"id":"notionTestId","name":"Notion Test","picture":"http://example.com/avatar","verified_email":true}}`
+)
+
+func (ts *ExternalTestSuite) TestSignupExternalNotion() {
+	req := httptest.NewRequest(http.MethodGet, "http://localhost/authorize?provider=notion", nil)
+	w := httptest.NewRecorder()
+	ts.API.handler.ServeHTTP(w, req)
+	ts.Require().Equal(http.StatusFound, w.Code)
+	u, err := url.Parse(w.Header().Get("Location"))
+	ts.Require().NoError(err, "redirect url parse failed")
+	q := u.Query()
+	ts.Equal(ts.Config.External.Notion.RedirectURI, q.Get("redirect_uri"))
+	ts.Equal(ts.Config.External.Notion.ClientID, q.Get("client_id"))
+	ts.Equal("code", q.Get("response_type"))
+
+	claims := ExternalProviderClaims{}
+	p := jwt.Parser{ValidMethods: []string{jwt.SigningMethodHS256.Name}}
+	_, err = p.ParseWithClaims(q.Get("state"), &claims, func(token *jwt.Token) (interface{}, error) {
+		return []byte(ts.Config.JWT.Secret), nil
+	})
+	ts.Require().NoError(err)
+
+	ts.Equal("notion", claims.Provider)
+	ts.Equal(ts.Config.SiteURL, claims.SiteURL)
+}
+
+func NotionTestSignupSetup(ts *ExternalTestSuite, tokenCount *int, userCount *int, code string, user string) *httptest.Server {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.URL.Path {
+		case "/v1/oauth/token":
+			*tokenCount++
+			ts.Equal(code, r.FormValue("code"))
+			ts.Equal("authorization_code", r.FormValue("grant_type"))
+			ts.Equal(ts.Config.External.Notion.RedirectURI, r.FormValue("redirect_uri"))
+
+			w.Header().Add("Content-Type", "application/json")
+			fmt.Fprint(w, `{"access_token":"notion_token","expires_in":100000}`)
```

Comment:
sanity check here - should match the format of the api response from notion's `/v1/oauth/token` endpoint